### PR TITLE
Feature/add tags to recent files

### PR DESCRIPTION
**Pull request by danielronnkvist**
Functionality for adding tags to newly added files.
When files have been chosen and save is clicked, the array is passed via the _$rootScope_ to a new Angular controller.

### CODE REVIEW

**Comment by nattravn** [comment]
I did not understand much of the Angularjs part, but everything else looked good. We can probably delete the file views / tags / new.html.erb

**Comment by klaseskilson** [comment]
I don't think we should delete `views/tags/new.html.erb`, since that is the only way to add tags to files after they have been added.

**Comment by nattravn** [comment]
True! ^^

File: app/assets/javascripts/controllers/TagCtrl.coffee [hunk 1]
```diff
@@ -0,0 +1,26 @@
+@bruseApp.controller 'TagCtrl', ['$scope', '$rootScope', '$location', '$routeParams', 'TagHandler', ($scope, $rootScope, $location, $routeParams, TagHandler) ->
+  if !$rootScope.new_files || !$rootScope.new_files.length
+    $location.path('/service/'+$routeParams.identity_id+'/files/add')
+
+  $scope.files = $rootScope.new_files
+  $scope.tags = ""
+
+  $scope.save = ()->
```
Comment (klaseskilson) [inline app/assets/javascripts/controllers/TagCtrl.coffee]: (Minor bullshit comment: [No empty parameter lists!](https://github.com/styleguide/javascript))

File: app/assets/javascripts/controllers/TagCtrl.coffee [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    success = true
+    tagsForAll = _.compact $scope.tags.split ' '
+    ###
+    For each file, split the input field by space into an array.
+    Remove empty strings with the _.compact method.
+    And lastly combine two arrays with concat.
+    ###
+    $scope.files.forEach((file)->
```
Comment (klaseskilson) [inline app/assets/javascripts/controllers/TagCtrl.coffee]: (Minor bullshit comment number two: [Spaces before function arrows!](https://github.com/styleguide/javascript))

File: app/assets/javascripts/controllers/TagCtrl.coffee [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      file.tags = "" if !file.tags
+      tagsForFile = _.compact file.tags.split ' '
+      file.new_tags = tagsForFile.concat tagsForAll
+      if file.new_tags.length > 0
+        TagHandler.put(file.bruse_file.id, file.new_tags).then((data)->
+          success = data.success
+          )
+      )
+    window.location.replace('/bruse_files') if success
```
Comment (klaseskilson) [inline app/assets/javascripts/controllers/TagCtrl.coffee]: Isn't there a risk that this happens before `TagHandler.put(...` is finished? JS won't wait for the asynchronous call to the backend.

**Comment by klaseskilson** [comment]
Ruby filenames should be snake_cased to follow the [style guide](https://github.com/bbatsov/ruby-style-guide#naming) on which GitHub's style guide is based.

**Comment by klaseskilson** [comment]
I'm not sure I understand what gets prioritized by the system here. If I go to `/service/:identity_id/files/add`, where does it take me? Angular or Rails?

File: app/assets/javascripts/controllers/TagCtrl.coffee (hunk 3 above)
Comment (danielronnkvist) [inline app/assets/javascripts/controllers/TagCtrl.coffee]: I could put the loop logic in the TagHandler instead, and then call the redirect in the callback.

**Comment by danielronnkvist** [comment]
`/service/:identity_id/files/add` is redirected to the `views/files/files/new.html.erb`. And that's where Angular takes over the control of the request.

**Comment by klaseskilson** [comment]
Ok. So [this](https://github.com/tkomstadius/bruse/blob/69986ea23b817c7b4c835e1710a7919a2accd843/app/assets/javascripts/main.coffee#L5-L9) get triggered only then? Or why is there duplicate route logic?

**Comment by danielronnkvist** [comment]
No. [This](https://github.com/tkomstadius/bruse/blob/69986ea23b817c7b4c835e1710a7919a2accd843/app/assets/javascripts/main.coffee#L10-L14) gets triggered. Since the `<base href="/">` tag is changing the base url of the request to only `/`. I think we have to use duplicate logic to make Rails routing work with Angular routing.

And since the Angular routing is in HTML5 mode the `/service/:identity_id/files/add/tag` renderes the add tags template. But in the `TagsCtrl.coffee`-file I check if their are any files on the `$rootScope`. If not the user is redirected to `/service/:identity_id/files/add`.

Think of it like this; once Rails have rendered the `app/views/files/files/new.html.erb`, Angular takes over the routing control.

**Comment by klaseskilson** [comment]
All right! Cool.

**Comment by danielronnkvist** [comment]
I have fixed the issue with asynchronous javascript. But we have a problem with all the request that are being sent to the server. If I add approx. 200 files, and then add tags to all the files, the system is still working on the add files process, so the request to add tags is delayed.

This problem is probably an effect of the _Fuzzily_ gem, if could postpone the creation of all trigrams I think the problem with the delay will be solved.

**Comment by klaseskilson** [comment]
What if we create a bulk method of some sort, so that we can decrease the number of http calls?

So instead of 200 `POST /service/123/files` calls, we use one `POST /service/123/bulk` call.

**Comment by klaseskilson** [comment]
Or investigate if the after_save (or whatever hook is beeing used) in Fuzzily can be prevented, and use some sort of delay job?

**Comment by danielronnkvist** [comment]
We could do that, but would that solve the problem with how much time Fuzzily takes? There is a method in Fuzzily to create trigrams after an object have been created, so a delayed job is possible.

**Comment by klaseskilson** [comment]
What if we wrap everything [`ActiveRecord::Transactions`](http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html) then? This would allow the server to only do one db commit (which is basically what takes time) instead of one for each trigram.

This probably has some disadvantages though. For one, there might be a maximum number of SQL commands per commit. I also think that if one query fails, it leads to a total collapse of the entire transaction (:poop:). [Interesting read on the subject!](https://www.coffeepowered.net/2009/01/23/mass-inserting-data-in-rails-without-killing-your-performance/)

Like this:

``` ruby
ActiveRecord::Base.transaction do
  1000.times { Model.create(options) }
end
```

**Comment by danielronnkvist** [comment]
A bulk request to the server that will be handled by the `ActiveRecord::Transactions` sounds good. Screw failures, speed over reliability hehe :smirk_cat:

**Comment by klaseskilson** [comment]
Looks good to me!

**Comment by nattravn** [comment]
The upload function for local files doesn't work now but maybe that isn't a problem if we gonna change the structure in the feature

**Comment by klaseskilson** [comment]
I don't think we should approve something that's broken to be merge into develop. Sure, we might change how things work down the line, but that doesn't change what we have today.

**Comment by danielronnkvist** [comment]
But is it the file upload or the function to add tags in post that is broken?

**Comment by nattravn** [comment]
The upload function/button doesn't work for me anymore. Byebug says that the identity_id from the query is null  ....AND "bruse_files"."identity_id" IS NULL) LIMIT 1. If the identity_id is null it throwing you to the else-state "you must log in to your bruse acount!" and I signed in as a bruse user

**Comment by nattravn** [comment]
I had to reset the database and that fixed my problem (I always forget to do that). But the add tag function doesn't work

**Comment by danielronnkvist** [comment]
I have fixed the issue with adding tags in post.

I have another solution to the angular routing problem. Instead of having the controller in `otherwise`, we could set all links in the header with the HTML attribute `target="_self"`. So it would look something like this:

``` ruby
<%= link_to 'My files', bruse_files_path, target: "_self"  %>
```

By doing like that the browser is forced to call the server and Angular wont intercept. It will also prevent a small loading time caused by the redirect controller.

**Comment by klaseskilson** [comment]
I think it's ready for merge!

**Comment by nattravn** [comment]
It works very good! I merge it

**Comment by johanhenrikssn** [comment]
Amazing!

**Pull request merged**